### DISCUSSION TITLE
How to support json-serialization for non-pydantic classes

### DISCUSSION THREAD

**Discussion by a-recknagel** [pydantic/pydantic#2920] [Question] [answered]
I have a `pydantic.BaseModel` that I want to call `.json()` on. The issue is that one of its attributes is a custom class that can't inherit from `pydantic.BaseModel`. I tried to duck-type the class into a pydantic shape, but still can't get it to work:

```Python
from pydantic import BaseModel, create_model


class Foo:
    def __init__(self, x: int):
        self.x = x

    # need this to have a json-schema
    __pydantic_model__ = create_model("FooModel", x=(int, ...))

    # need this to be de-serializeable
    @classmethod
    def __get_validators__(cls):
        yield cls


class Response(BaseModel):
    foo: Foo

    class Config:
        json_encoders = {
            # do I really have to register that here, and not on `Foo` where it belongs?
            Foo: lambda foo: {"x": foo.x}
        }

Response(foo=Foo(1)).json()  # TypeError: Object of type Foo is not JSON serializable 
```

I'm reaching the limit of debugging my abilities here, because when I debug the `.json()` call it sends me straight into standard library modules `typing`, `abc` and (drumroll) `json`, where it ultimately fails. What's up with that anyway, the method is registered on `BaseModel` but not a single line of it gets executed when called, the whole thing is really opaque.

To address some plausible but not possible solutions:
 - I can't pass a custom encoder to the `.json()` call, because a framework does the call (fastapi, when turning a response object into json)
 - I can't let `Foo` inherit from `pydantic.Basemodel` because it's already the child of a 3rd party class, which gives it some attributes that I need to have in its json representation
 - I can't mirror these classes into an equivalent `pydantic` schema, because my real-life example has three levels of ~80 attributes, with documentation and all. I don't want to maintain all of that twice just because I can't get a single class to cooperate.

**Comment by PrettyWood** [comment, marked as answer]
Hi @a-recknagel 
I don't understand the problem. This works perfectly
```py
from pydantic import BaseModel


class Foo:
    def __init__(self, x: int):
        self.x = x


class Response(BaseModel):
    foo: Foo

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {
            Foo: lambda foo: {"x": foo.x}
        }


assert Response(foo=Foo(1)).json() == '{"foo": {"x": 1}}'
```

To answer the comment in the code yes for now you need to write it in the `Config` and not directly in the class itself.
You can always create a custom `BaseModel` or json encoder if you want to avoid copy paste!
FYI there is https://github.com/samuelcolvin/pydantic/issues/951 that may be tackled some day.

Hope it helps!

**Reply by a-recknagel** [reply]
You're right, that should teach me to not post questions (with snarky undertones, no less) if I'm too tired to even notice when my MCVE suddenly starts working =(

For posterity, this here is the full context of my problem, which also works out fine:

```Python
from fastapi import FastAPI
from pydantic import BaseModel, create_model

app = FastAPI()


class Foo:
    def __init__(self, x: int):
        self.x = x

    @classmethod
    def ctor(cls, v):
        """Is this how it's supposed to be done?"""
        if isinstance(v, cls):
            return v
        elif isinstance(v, dict):
            return cls(**v)
        else:
            raise TypeError(f"Type {type(v)} not accepted yet.")

    @classmethod
    def __get_validators__(cls):
        yield cls.ctor

    __pydantic_model__ = create_model(
        "FooModel",
        __config__=type("Config", (), {"title": "Foo"}),
        x=(int, ...)
    )


class Response(BaseModel):
    foo: Foo

    class Config:
        json_encoders = {
            Foo: lambda foo: {"x": foo.x}
        }


@app.get("/a", response_model=Response)
def from_dict():
    return {"foo" : {"x": 1}}


@app.get("/b", response_model=Response)
def from_obj():
    return Response(foo=Foo(1))


if __name__ == '__main__':
    import uvicorn
    uvicorn.run("main:app")
```

I need the `Config` in foo to get an OpenAPI representation, and the `__get_validators__` so that both building a response from an instance (route b) and from json (route a) do work.

**Reply by a-recknagel** [reply]
I reposted this QA to stackoverflow, mainly to increase visibility in case someone else runs into the same issue https://stackoverflow.com/questions/68016155/how-to-include-non-pydantic-classes-in-fastapi-responses/68016156#68016156

@PrettyWood If you find the time, it might be nice if you could also comment or improve the answer over there.
